src/Collection/Iterator/UniqueIterator.php

```diff
@@ -0,0 +1,78 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Collection\Iterator;
+
+class UniqueIterator extends FilterIterator
+{
+    protected array $_exists = [];
+
+    /**
+     * @param \Traversable|array $items The items to be filtered.
+     * @param string|null $key An object property or array key that will be used to determine uniqueness.
+     */
+    public function __construct($items, ?string $key)
+    {
+        parent::__construct($items, function ($item) use ($key) {
+
+            if (is_null($key)) {
+                $key = $value = $item;
+            } else {
+                if (is_object($item) && method_exists($item, 'get' . ucfirst($key))) {
+                    $value = $item->{'get' . ucfirst($key)};
+                } elseif (is_object($item) && method_exists($item, '__toString')) {
```

Comment:
Why is `__toString` preferred over accessing properties? This path should raise an error if `$key` is provided as we've given up trying to unique values based on the `key` parameter.